### DISCUSSION TITLE
is there charset code that support  Arabic text

### DISCUSSION THREAD

**Discussion by ebrahimmansur** [kakzaki/blue_thermal_printer#163] [General]
need to print invoice in Arabic...thank you

**Comment by kreemov** [comment]
windows-1256 worked with me

**Comment by ebrahimmansur** [comment]
please provide gist of the code ....here is mine 

` String voucherNameWithValueMessage = 'ليبيانا  5 دينار';
    blueThermalPrinter.printCustom(
        voucherNameWithValueMessage, _onlyBoldText, _alignCenterText,
        charset: 'windows-1256');`

**Comment by fatimaown93** [comment]
> please provide gist of the code ....here is mine
> 
> ` String voucherNameWithValueMessage = 'ليبيانا 5 دينار'; blueThermalPrinter.printCustom( voucherNameWithValueMessage, _onlyBoldText, _alignCenterText, charset: 'windows-1256');`

me too it's not working for me 
did you solve it ? if so 
I really need help

**Comment by fatimaown93** [comment]
> windows-1256 worked with me

how did you did it ?!

**Comment by ebrahimmansur** [comment]
still nothing....

**Comment by fatimaown93** [comment]
> still nothing....

actually I find a way but it's really twisted  🤣
![WhatsApp Image 2021-07-30 at 2 45 26 PM](https://user-images.githubusercontent.com/67221573/127648542-2a742f55-bb28-4318-b89d-9e9a949ecb24.jpeg)
I used canvas textpainter and then convert it to an image 
then I used the function printImageBytes 
It works but there was a problem the function printimage has a fixed height for the image [200px - 250px]. I will try to make it more than one image on sunday.

**Comment by ebrahimmansur** [comment]
> > please provide gist of the code ....here is mine
> > ` String voucherNameWithValueMessage = 'ليبيانا 5 دينار'; blueThermalPrinter.printCustom( voucherNameWithValueMessage, _onlyBoldText, _alignCenterText, charset: 'windows-1256');`
> 
> me too it's not working for me
> did you solve it ? if so
> I really need help



> > still nothing....
> 
> actually I find a way but it's really twisted 🤣
> ![WhatsApp Image 2021-07-30 at 2 45 26 PM](https://user-images.githubusercontent.com/67221573/127648542-2a742f55-bb28-4318-b89d-9e9a949ecb24.jpeg)
> I used canvas textpainter and then convert it to an image
> then I used the function printImageBytes
> It works but there was a problem the function printimage has a fixed height for the image [200px - 250px]. I will try to make it more than one image on sunday.

not an efficient solution ....i thought  about make it an pdf then printing it as an image before.....still not really a solution 😂...so still waiting on @kreemov to show me if it really works.

**Comment by Salakar** [comment]
What's the result of using `utf-8` as the `charset`?

**Comment by ebrahimmansur** [comment]
> What's the result of using `utf-8` as the `charset`?

this is the output

![IMG_20211203_212732](https://user-images.githubusercontent.com/34646327/144661869-022077c2-f076-4b2f-a553-c44ac6e09e8d.jpg)
.

**Comment by Salakar** [comment]
Can I ask the make/model of the printer?

**Comment by ebrahimmansur** [comment]
that is the problem we don’t have constraint on what printer the merchant
uses, we only provide the connection and printing functionality....if we
did we could have used the printer's sdk...😁

**Comment by Salakar** [comment]
I see 😅 the only other thing I can suggest is try print through a quick Node.js script with something like https://www.npmjs.com/package/node-thermal-printer that's known to support Arabic - if it works with that then it confirms it's an issue with this plugin and not the printer

**Comment by ebrahimmansur** [comment]
Thanks ...and also thanks a lot for the support ✌️ ... i appreciate the time .......can’t with for the next space ...it’s seems my issue with the "multi package"  was not fully understood because of its complexity and with out having visual representation 😅😭

**Comment by bluemix** [comment]
Go to the device POS settings, and on print settings, set printing with UTF-8.

If the above method doesn't work, try the next method (which is printing images as text).

Solution inspired from @fatimaown93.

You can convert the widget into an image (using [Screenshot)](https://pub.dev/packages/screenshot) and print it:

![image](https://user-images.githubusercontent.com/3332274/155361630-89177078-2600-4a26-8d84-3370a85b4ff4.png)


Here is the Widget:
```dart
Screenshot(
    controller: screenshotController,
    child: Container(
      width: 190,
      decoration: BoxDecoration(
          border: Border.all(color: Colors.black),
        color: Colors.white,
      ),
      // color: Colors.white,
      child: Column(
        children: [
          Row(
            children: [
              Text(
                'أهلاً وسهلاً',
                style: TextStyle(
                  color: Colors.black,
                ),
                textAlign: TextAlign.right,
              ),
              Spacer(),
              Text(
                'شكو ماكو؟',
                style: TextStyle(
                  color: Colors.black,
                ),
              ),
            ],
          ),
  
        ],
      ),
    ),
  ),
```

The widget is converted into image of `Uint8List`:

```dart
var imageTextRawUint8List = await screenshotController.capture() ?? Uint8List(0);
```

Then, print the image bytes:

```dart
 bluetooth.printImageBytes(imageTextRawUint8List);
```

**Comment by ebrahimmansur** [comment]
Thanks so much @bluemix  ,but I have already thought about it and implemented it if you have seen the conversion between me and @fatimaown93, but I want is an efficient solution , I  wish if there was a way to convert codepage to utf-8 command as binary message to the thermal printer.  Im still trying to test some things , I will keep you posted if I came with a better solution (I don't want to depend on a package not made for this problem and bound/scoped to a UI logic that can change over time ) .

**Comment by iyashiyas** [comment]
> windows-1256

how? can u share the code please 
or correct if i am wrong
 Uint8List encoded =
              await CharsetConverter.encode('windows-1256', message.toString());
              printer.textEncoded(encoded);

**Comment by bluemix** [comment]
you have to go the printer settings menu (Android settings) and set the encoding to utf-8 or windows-1256

**Comment by devmuaz** [comment]
Changing the `printer settings (can be found in the android settings)` worked for me, I set it to `UTF-8` and now it's printing Arabic characters!
Thanks @bluemix 👏

**Comment by blitzlepe** [comment]
> Can you share the path of file(to) to edit?



> Changing the `printer settings (can be found in the android settings)` worked for me, I set it to `UTF-8` and now it's printing Arabic characters! Thanks @bluemix 👏

Can you share the path of file(to) to edit?

**Comment by devmuaz** [comment]
@blitzlepe It's in the system settings, there must be a section called **Printer** where you can change the encoding to UTF-8

**Comment by faizanShoukat55** [comment]
Thank you @bluemix  you saved my day. I was trying it from the last few days.
